Added a Whipple model example that needs a little love.

moorepants opened the pull request:
Just moving this file that was in `pydy/codegen/tests/whipple.py` to a new branch since it doesn't quite work.
- [ ] There are no merge conflicts.
- [ ] A figure (preferrably SVG) has been added for the example.
- [ ] The required versions of all dependencie have been noted. (For
  notebooks, the [version information
  extension](https://github.com/jrjohansson/version_information) is helpful
  for this.)
- [ ] A `run.py` file that executes the example is included.
- [ ] All reviewer comments have been addressed.

# Task list

- [x] My results using floating point operations do not match the Basu-Mandal numbers to machine precision.
  - Let's open a separate issue about this because it will take some careful look at how the equations of motion are formed and why the specific encoded arithmetic is susceptible to floating point error.
- [ ] Determine if the numerical results differ with and without cse'ing the expressions.
  - [x] Check by setting cse=True/False in CythonMatrixGenerator or CythonODEGenerator.
  - [ ] Check by lambdifying and evaluating subexpressions and comparing to lambdifying and evaluating the whole expression.
  - [x] Check by xreplacing and evaluating subexpressions and comparing to xreplacing and evaluating the whole expression.
- [ ] Determine if and why there are differences in numerical evaluation with floats using `subs()`, `xreplace()`, `autowrap()`, `lambdify()`, `CythonMatrixGenerator()`, `LambdifyMatrixGenerator()`, `TheanoMatrixGenerator()`.
- [x] Independently calculate Kane's equations and compare to results of KanesMethod.
- [ ] Try out the csejac from https://github.com/sympy/sympy/issues/8191
- [x] Make sure to set the sympy cache value to a high number as it may speed up many operations, also see https://github.com/sympy/sympy/issues/8191.
  - The cache has been set to 6000 and the beginning of the script.
- [x] Check whether solving u' = inv(M)*F symbolically removes the floating point accuracy issues.
  - Floating point errors from the cse'd C code based on symbolic udots are present and it takes 12+ hours to compile. Solving symbolically does not resolve the floating point error issue.
- [x] Use symengine to make the reference calculations for numerical accuracy.
  - Using symengine to evaluate the mass matrix and forcing vector to 18 decimal place precision and then using np.linalg.solve() to find the state derivatives of the dynamical equations give results that match the Basu-Mandal numbers to machine precision. This verifies that the symbolic equations of motion are correct.
- [x] Try cse'ing the mass matrix and forcing vector, then solving the simplified linear system. That result could then be cse'd too.
  - It works! And is very fast for this system. PR #451 implements it.
- [x] Figure out why xdd and ydd are incorrect. There must be an error in the calculations, as all others match to machine precision with the symengine computed values.
  - I incorrectly defined u1 and u2 by not including the effects of u5. The wheel velocity is relative to the rear frame C, so both u5 and u6 contribute to the ground path velocity.
- [ ] See what the effects are of setting assumptions on the constants: nonnegative, real, etc.
- [ ] Using symengine to precisely evaluate the M and F from KanesMethod() and my independent formulation give -2 as the relative error between the results of the dynamics equations. Investigate this in the checks.py file.
- [ ] Try using SymPy's SymbolicSystem to return the relevant symbolic information from the derivation.
- [ ] I don't think `KanesMethod()` solves for the dependent generalized speeds. It simply differentiates the nonholonomic constraint equations and then the derivative's of the dependent generalized speeds are solved for by either `KanesMethod.rhs()` symbolically or with the `ODEFunctionGenerator()` numerically. This means that the dependent generalized speeds used in evaluating the equations of motion during simulation are from numerically intergration from the prior timestep or from the initial conditions. We set the roll rate initial condition in the simulation, but do not calculate the dependent speeds initial conditions using the nonholonomic constraints, so that could be an initial condition error. Secondly, it would be much better if the nonholonomic constraints are actually used to calculate the equations of motion and only based on the independent speeds and coordinates. The generated c code would be better if it followed this form:
  ```python
   def eval_full_rhs(u_ind, q_ind, t):
      # this would solve for the pitch angle using fsolve internally (or don't do this and use a DAE integrator)
      q_dep = solve_dep_coordinates(q_ind, q_dep_guess, t)
      # this would calculate u3, u5, u8 exactly given u4, u6, u7 and the q's
      u_dep = eval_dep_speeds(u_ind, q_ind, q_dep, t)
      # this gives all qdots from q3' to q8'
      q_dots = eval_kin_diff_eqs(u_ind, u_dep, q_ind, q_dep, t)
      # this gives all udots from u3' to u8'
      u_dots = eval_udots(u_ind, u_dep, q_ind, q_dep, t)
      return q_dots, u_dots
  ```

# Things learned

- With no assumptions applied to any symbols and using KanesMethod to arrive at the equations of motion, the mass matrix has 1,173,655 operations (takes 10.5 s to count_ops) and the forcing vector has 473,692 operations (takes 2.28 s to count_ops).
- If you cse the mass matrix and forcing vector together (from KanesMethod) there are 577 common subexpressions that result in 2367 total operations. That is a 99.86% reduction in the number of operations.
- If you cse the result of `kane.rhs()` it takes a long time to complete. I haven't waited it out yet.
- It took over 12 hours to run CythonODEFunctionGenerator() (with cse) on `kane.rhs()`. Not sure what the slowest parts are, but cse of the rhs is slow (see above). kane.rhs() is quite fast.
- If you count_ops the result of `kane.rhs()` it takes a long time to complete. I haven't waited it out yet.
- The mass matrix will have significantly more operations if real, positive, or nonnegative assumptions are placed on the constants, than if no assumptions are made on any of the constants.
- Using `.evalf(subs={})` to numerically evaluate the mass matrix or forcing vector with float substitutions takes over 3 hours to complete on my laptop (3+ hrs for each).
- Using `lambdify()` on the mass_matrix will kill the python interpreter.
- `.xreplace()` on the mass matrix seems to be the only SymPy way to evaluate with floats in a reasonable amount of time.
- Comparing `CythonMatrixGenerator()` evaluations of the mass matrix with floats using cse and not using cse results in different results. The no cse version takes some time for gcc to compile. The third row of the mass matrix has errors from 1e-6 to 1e-3. All other rows have errors less than 1e-14. The third row corresponds to u7, the steer generalized speed. I think uses like 10 Gb of data when compiling without cse.
- gcc can't seem to handle CythonMatrixGenerator of both the mass matrix and forcing vector at the same time without cse. You have to do them separately.
- autowrap() of the mass matrix works (no cse is used in autowrap) for both C and Fortran. The resulting numerical values are identical.
- Using floats with `xreplace()` and `autowrap()` give different results with the mass matrix, meaning they must evaluate the expressions with a different order of operations.
- Both C and Fortran backends for autowrap() give the same numerical results, meaning they generate the same order of operations.
- It may be useful to raise the sympy cache size for the differentiation

# Issues to open on SymPy

- [ ] find_dynamicsymbols should deal with Vectors and lists of exprs/vectors/etc.
- It would be convenient if this worked:
  ```python
  N = ReferenceFrame('N')
  A = ReferenceFrame('A')
  A.orient(N, 'Axis', (10, A['3']))
  ```
  But you have to use `N['3']` in the last line.
- cse() takes forever on results of `kane.rhs()`
- count_ops, find_dynamicsymbols, free_symbols can be slow for large expressions
- xreplace() doesn't necessarily give same results as autowrap() due to different order of operation evaluations
- autowrap() can't work with Function('q')(t) like lambdify does
- raise an issue about not knowing which order the x vector is in with reference to M * x' = F (like what order is mass_matrix_full in?)


# Notes

- Mapping of my coordinates to Gilbert's in the SymPy documentation:
  ```
  # var   : J , G
  # yaw   : q3, q1
  # roll  : q4, q2
  # pitch : q5, q4
  # steer : q7, q5
  # rwh   : u6, u3
  # fwh   : u8, u6
  ```

jcrist commented:
+1

moorepants commented:
I've been doing some work on this. Haven't yet quite got it where I want it.

moorepants commented:
@pydy/pydy-developers 

I've almost got this example working. There are two issues (possible related) that I could use some more eyes helping me spot the bug:

1. I get different results for the state derivatives when substituting values into the expressions versus using the generated right hand side.
2. The substitution results are close to the correct values, but not quite. There must be a little error somewhere.

Current results on my machine:
```
(pydy-dev-examples)moorepants@moorepants-2170p:bicycle(whipple)$ ipython --pdb
Python 2.7.9 |Continuum Analytics, Inc.| (default, Apr 14 2015, 12:54:25) 
Type "copyright", "credits" or "license" for more information.

IPython 3.1.0 -- An enhanced Interactive Python.
Anaconda is brought to you by Continuum Analytics.
Please check out: http://continuum.io/thanks and https://binstar.org
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.

In [1]: run whipple.py
Defining reference frames.
Defining time varying symbols.
Orienting frames.
Defining constants.
Defining position vectors.
Defining holonomic constraints.
The holonomic constraint is a function of these dynamice variables:
[q4(t), q5(t), q7(t)]
Defining kinematical differential equations.
Defining angular velocities.
Defining linear velocities.
Defining nonholonomic constraints.
The nonholonomic constraints are a function of these dynamice variables:
[q4(t), q5(t), q7(t), u3(t), u4(t), u5(t), u6(t), u7(t), u8(t)]
Defining inertia.
Defining the rigid bodies.
Defining the forces and torques.
Generating Kane's equations.
The mass matrix is a function of these dynamic variables:
[q4(t), q5(t), q7(t)]
The forcing vector is a function of these dynamic variables:
[T4(t), T6(t), T7(t), q4(t), q5(t), q7(t), u3(t), u4(t), u5(t), u6(t), u7(t), u8(t)]
Loading numerical input parameters.
ie23 not added to sub dict.
ic12 not added to sub dict.
ie12 not added to sub dict.
id33 not added to sub dict.
ic23 not added to sub dict.
if33 not added to sub dict.
Substituting numerical parameters into SymPy expressions.
The state derivatives from substitution:
[ 7.86046362 -1.86557821  4.53122639 -0.8514172  -0.11319592 -2.47942289]
Generating a right hand side function.
Generating rhs
The state derivatives from code gen:
[-0.78300335 -0.60684258 -0.48598247  0.01191855  7.85918311 -1.84409601
  4.54243564 -0.84781901 -0.11415563 -2.45917668]
Generating output dictionary.
Assertions.
Failed: psifdd:
  Correct: -4.6198904039402997
Incorrect: -4.5312263892435896
Failed: psifdd:
  Correct: -4.6198904039402997
Incorrect: 0.4859824687093000
Failed: xdd:
  Correct: -0.5041626315047000
Incorrect: -0.4763002611500358
Failed: xdd:
  Correct: -0.5041626315047000
Incorrect: 0.0519495313636576
Failed: ydd:
  Correct: -0.3449706619454000
Incorrect: -0.3465658441625137
Failed: ydd:
  Correct: -0.3449706619454000
Incorrect: -2.4559798471744747
Failed: psidd:
  Correct: -7.8555281128244001
Incorrect: -7.8604636243741410
Failed: psidd:
  Correct: -7.8555281128244001
Incorrect: 0.7830033527065000
Failed: thetadd:
  Correct: 0.8353281706379000
Incorrect: 0.8514172038935794
Failed: thetadd:
  Correct: 0.8353281706379000
Incorrect: -0.0119185528069000
Failed: phidd:
  Correct: 0.1205543897884000
Incorrect: 0.1131959163667986
Failed: phidd:
  Correct: 0.1205543897884000
Incorrect: -7.8591831087039372
Failed: zdd:
  Correct: -1.4604528332980000
Incorrect: -1.4613139486634557
Failed: zdd:
  Correct: -1.4604528332980000
Incorrect: 0.0467409205129194
Failed: betafdd:
  Correct: 2.4548072904549998
Incorrect: 2.4794228928134645
Failed: betafdd:
  Correct: 2.4548072904549998
Incorrect: 1.8440960096057826
Failed: betardd:
  Correct: 1.8472554144216999
Incorrect: 1.8655782098325024
Failed: betardd:
  Correct: 1.8472554144216999
Incorrect: 0.6068425835418000
```

moorepants commented:
@chrisdembia Would you have a few minutes to look this over? You should be able to run it with PyDy master.

oliverlee commented:
@moorepants Do you think it'd take a lot of work to have dtk support Python 3?

moorepants commented:
No it should be minimal. We could just copy over the `dtk.bicycle` module (or relevant functions) to here too. I just have all the basumandall stuff coded up there as I used it in my dissertation to check my old autolev model.

oliverlee commented:
Issue 1 fixed here: https://github.com/oliverlee/pydy/commit/bc21eb2b6bef29a3317675ee4cd4070344eeca56

moorepants commented:
That's awesome! So there is now just something minor (and probably even harder to find...).

Do you want to make a PR against this branch?

oliverlee commented:
I can do that or you call pull in my changes if you think you'll look at it
soon.
On Aug 19, 2015 19:19, "Jason K. Moore" <notifications@github.com> wrote:

> That's awesome! So there is now just something minor (and probably even
> harder to find...).
>
> Do you want to make a PR against this branch?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pydy/pydy/pull/122#issuecomment-132700476>.
>

moorepants commented:
Yeh, I can just pull your branch and merge.

moorepants commented:
Note that I was able to get very high precision matching with my old Autolev code: http://moorepants.github.io/dissertation/eom.html#non-linear-validation, so I'm inclined to believe that this model still has a small issue, but we also can't rule out that the double precision evaluation here isn't the same as what I got with the code generated by Autolev.

moorepants commented on examples/bicycle/whipple.py:
```diff
@@ -0,0 +1,508 @@
+####################################
+# Validation of non-linear equations
+####################################
+
```
One thing to check: how am I handling the holonomic constraint in the numerical evaluation? It isn't clear if I am.

moorepants commented:
These two don't seem to give the same results for some reason. The lower one caused the results in the manual kanes equations functions to give slightly different final xd results than the equations derived from KanesMethod class.
```
@@ -193,29 +135,19 @@ def formulate_nonmin_equations_motion(newtonian_frame,
     kin_diff_map = solve_for_qdots(q, generalized_speed_defs)
 
     print('Solving for the dependent generalized speeds')
-    A_GuD, B_G = decompose_linear_parts(motion_constraints, uD)
-    uD_of_uI = A_GuD.LUsolve(-B_G)
+    A_GuI, A_GuD, B_G = decompose_linear_parts(motion_constraints, uI, uD)
+    uD_of_uI = A_GuD.LUsolve(-A_GuI*sm.Matrix(uI) - B_G)
```

moorepants committed:
Remove the documentation example from this PR.

moorepants committed:
Added a test for compare_cse.

moorepants committed:
Use minimal args in lambdify calls in compare_cse.

moorepants committed:
Trying compare_cse on the mass matrix.

moorepants committed:
Python interpreter was getting killed when lambdifying the mass matrix it seems.

moorepants commented:
I've noticed that if I use `0.0`, `0`, or `sympy.S(0)` for the zero quantities when setting velocities to zero and the inertia scalars to zero that I get different results in the simulation with all three! Something fishy there.

moorepants committed:
Try xreplace instead of evalf.

moorepants committed:
Fix numpy to numpy comparison, F works and M does too but has some 0 to 0 comparisons.

moorepants committed:
Added some docs to compare_cse and removed print statements. [skip ci]

moorepants commented:
Oscar shared this on the SymPy list that could help with the cse comparisons:

```python
rom sympy.utilities.iterables import postorder_traversal

def cse_compare(expr, syms, vals):
    result = lambdify(syms, expr)(*vals)

    syms_all, vals_all = list(syms), list(vals)
    reps, [reduced_expr] = cse(expr)
    for csesym, cseexpr in reps:
        val = lambdify(syms_all, cseexpr)(*vals_all)
        syms_all.append(csesym)
        vals_all.append(val)
    result_cse = lambdify(syms_all, reduced_expr)(*vals_all)

    return abs(result - result_cse) / abs(result)

def cse_compare_bottom_up(expr):
    syms = list(expr.free_symbols)
    vals = list(np.random.random(len(syms)))

    def subexprs():
        seen = set()
        for e in postorder_traversal(expr):
            if e not in seen:
                yield e
                seen.add(e)

    # possibly better:
    #exprs = sorted(set(postorder_traversal(expr)), key=lambda e: e.count_ops())
    exprs = subexprs()

    for subexpr in exprs:
        print()
        print(subexpr)
        print(cse_compare(expr, syms, vals))

test_expr = (sin(x+1) + 2*cos(x-3)).expand(trig=True)
cse_compare_bottom_up(test_expr)
```

moorepants committed:
Comparing C mass matrix and forcing evals. Getting different results.

moorepants commented:
With commit ef6a9e7 it's showing that the C code generated is not giving machine precision match on the evaluation of the mass matrix and forcing vector. The follow example shows errors as large as 1e-3 in some entries. (still doesn't explain not matching Basu-Mandal though).
```
ipdb> F1 - np.squeeze(F3)
array([-3.41060513e-13,  1.73117225e-03, -3.38840901e-02,  5.55111512e-17,
       -2.22044605e-16, -6.93889390e-17])
ipdb> M1.reshape((6, 6)) - M3
array([[-1.42108547e-14,  6.66133815e-16, -2.77555756e-17,
        -7.10542736e-15,  1.77635684e-15,  1.38777878e-17],
       [ 7.81242924e-05, -9.52409434e-05,  9.42249951e-07,
        -1.81921835e-04, -3.76608293e-04, -1.11431143e-06],
       [-1.52912027e-03,  1.86414305e-03, -1.84425798e-05,
         3.56074091e-03,  7.37132274e-03,  2.18103248e-05],
       [ 1.38777878e-17,  0.00000000e+00,  0.00000000e+00,
        -1.38777878e-17,  7.97972799e-17,  0.00000000e+00],
       [-3.64291930e-17,  0.00000000e+00,  0.00000000e+00,
        -1.11022302e-16,  1.11022302e-16,  1.38777878e-17],
       [ 1.38777878e-17,  0.00000000e+00,  0.00000000e+00,
         1.38777878e-17,  0.00000000e+00, -1.38777878e-17]])
```

moorepants committed:
Testing the cython matrix code with the mass matrix.

moorepants committed:
Added autowrap testing.

moorepants committed:
Added function to evaluate to specific precision with symengine and function to compare CythonMatrixGenerator with and without cse.

moorepants committed:
Added calls to the function to compare with and without cse.

moorepants committed:
Added a utility function for evaluating autowrapped functions and renamed the long function name.

moorepants commented:
It now seems clear that the floating point errors are causing the larger differences with the Basu-Mandal numbers. Using symengine to evaluate the expressions at an exact precision gives the matches, verifying the model is correct. Holly shit that was a rabbit hole. I still don't know why the derivation methods we employ for the EoMs here differ from autolevs. Here is a comparison I'm getting:

```
In [3]: basu_output
Out[3]: 
{'xdd': -0.5041626315047,
 'ydd': -0.3449706619454,
 'zdd': -1.460452833298,
 'thetadd': 0.8353281706379,
 'psidd': -7.8555281128244,
 'phidd': 0.1205543897884,
 'psifdd': -4.6198904039403,
 'betardd': 1.8472554144217,
 'betafdd': 2.454807290455}

In [4]: moore_output_basu_from_sub
Out[4]: 
{'x': 0.0,
 'y': 0.0,
 'z': 0.2440472102925096,
 'theta': 0.0,
 'psi': 0.9501292851472,
 'phi': 3.1257073014894,
 'betar': 0.0,
 'psif': 0.2311385135743,
 'betaf': 0.0,
 'xd': -2.8069345714545,
 'yd': -0.14809823960010002,
 'zd': 0.10587787462605407,
 'thetad': 0.7830033527065,
 'psid': 0.6068425835418,
 'phid': -0.0119185528069,
 'betard': 8.912989661489,
 'psifd': 0.4859824687093,
 'betafd': 8.0133620584155,
 'xdd': -0.467996314568172,
 'ydd': -0.347770341987600,
 'zdd': -1.460452833298032,
 'thetadd': 0.8353281706376864,
 'psidd': -7.855528112824378,
 'phidd': 0.12055438978862705,
 'betardd': 1.847255414421868,
 'psifdd': -4.6198904039391,
 'betafdd': 2.4548072904552583}
```

the xdd and ydd are off. I think those are incorrectly defined in this script, but the others match to machine precision.

moorepants commented:
I solved for the udots symbolically (LUSolve of the 6x6 matrix) and then CythonODEFunctionGenerated that. It took like 12+ hours to compile, which is nutz. I am not sure why this would be so different than the autolev version. Here is the timeit of the resulting compiled function:
```
In [2]: %timeit rhs(state_vals, 0.0, specified_vals, constants_vals)
8.34 µs ± 232 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```
And here are the results compared to the expected:
```
In [3]: basu_output
Out[3]: 
{'xdd': -0.5041626315047,
 'ydd': -0.3449706619454,
 'zdd': -1.460452833298,
 'thetadd': 0.8353281706379,
 'psidd': -7.8555281128244,
 'phidd': 0.1205543897884,
 'psifdd': -4.6198904039403,
 'betardd': 1.8472554144217,
 'betafdd': 2.454807290455}

In [4]: moore_output_basu_from_gen
Out[4]: 
{'x': 0.0,
 'y': 0.0,
 'z': 0.2440472102925096,
 'theta': 0.0,
 'psi': 0.9501292851472,
 'phi': 3.1257073014894,
 'betar': 0.0,
 'psif': 0.2311385135743,
 'betaf': 0.0,
 'xd': -2.8069345714545,
 'yd': -0.14809823960010002,
 'zd': 0.10587787462605407,
 'thetad': 0.7830033527065,
 'psid': 0.6068425835418,
 'phid': -0.0119185528069,
 'betard': 8.912989661489,
 'psifd': 0.4859824687093,
 'betafd': 8.0133620584155,
 'xdd': -0.475600581242521,
 'ydd': -0.347056598638782,
 'zdd': -1.4609631001925298,
 'thetadd': 0.8445459095327956,
 'psidd': -7.858452724615525,
 'phidd': 0.11671297721390103,
 'betardd': 1.8672421365703233,
 'psifdd': -4.573798129990226,
 'betafdd': 2.476956437395924}
```
Ignorning xdd and ydd because they likely have different errors, 

```
Failed: zdd:
  Correct: -1.4604528332980000
Incorrect: -1.4609631001925298
Failed: thetadd:
  Correct: 0.8353281706379000
Incorrect: 0.8445459095327956
Failed: psidd:
  Correct: -7.8555281128244001
Incorrect: -7.8584527246155247
Failed: phidd:
  Correct: 0.1205543897884000
Incorrect: 0.1167129772139010
Failed: psifdd:
  Correct: -4.6198904039402997
Incorrect: -4.5737981299902257
Failed: betardd:
  Correct: 1.8472554144216999
Incorrect: 1.8672421365703233
Failed: betafdd:
  Correct: 2.4548072904549998
Incorrect: 2.4769564373959239
```
The floating point errors are still present. In my old autolev code I do solve for the udots symbolically. And this shows that that isn't the issue.

moorepants committed:
With symengine arbitrary precision the model matches Basu-Mandall!

moorepants committed:
Added some documentation and return same shape array as matrix.

moorepants committed:
The symengine calculated values show the model matches Basu-Mandal.

moorepants committed:
Corrected the u1p, u2p, xdd, and ydd calcualtions (forgot to include effects of u5) and various other code cleanups.

moorepants committed:
Fixed merge conflict.

moorepants committed:
The new linear_sys_solver='sympy' builds and compiles super fast on the whipple model.

moorepants committed:
Removed some notes and added a data structure to contain the symbolics.

moorepants committed:
Moved symbolic derviation into a function.

moorepants committed:
Moved custom ReferenceFrame class into utils.py.

moorepants committed:
Added a function for the pretty printed Basu-Mandal value comparison.

moorepants committed:
Moved lots of code from whipple.py to functions in checks.py and utils.py.

moorepants committed:
Don't introduce both no and dn.

moorepants committed:
Don't delete the figures!

moorepants committed:
Clean up the whipple derivation a bit with formatting and notes.

moorepants committed:
Added a test for decompose_linear_parts that I thought would fail based on an error in the tire force branch.

moorepants committed:
Tried to ensure nonholonomic constraints hold on values passed into ode function, but it doesn't improve the result.

moorepants committed:
Swap order of nonholonomic constraint to fix numerical benchmark comparison.

Moved the derivative of the pitch holonomic constraint between the two
side slip nonholonomic constraints and the floating point comparison to
Basu-Mandal now works to machine precision. Thanks to Timo Stienstra for
discovering this. This is still a bug, but we are addressing it in
SymPy. For now, this at least verifies this model can evaluate correctly
with floats.

This PR was opened Feb 7, 2015 and now on March 3, 2023 the comparison
finally works. What an ordeal! But still more work to do, as solving
linear systems symbolically is not reliable for floating point
calculations. The divide-by-zero issues are caused by solving the
nonholonomic constraint equations symbolically internal to KanesMethod.

moorepants commented:
Couple of updates on this:

1. Solving nonholonomic constraints for the dependent generalized speeds symbolically can result in expressions that are not numerically stable (potentially both causing divide-by-zero issues and worse floating point error accumulation). `KanesMethod` solves these equations symbolically internally and the numerical comparison issues that have plagued this PR are most certainly due to this symbolic linear solve.
2. The symbolic linear solve in this problem is sensitive to the order of the nonholonomic constraint expressions. By swapping the order of the constraints as in https://github.com/pydy/pydy/pull/122/commits/02f23ae65f109d9da607055c5078f4d64909a184, the comparison to the Basu-Mandal numbers finally works (but this is only for that set of numerical inputs). I also tried the order `A['3'],A['1'],A['2']` but that was giving wildly different numerical results. I'm not sure why.

Thanks very much to @TJStienstra for discovering these details. Hopefully we can path SymPy to eliminate this source of error.

Peter230655 commented:
As the sequence of the entries in the speed_constraints list matters a lot, is it better to have the 'smaller' terms at the 'left' of the list ( small indices) or towards the end of the list (large indices)?
Thanks!

moorepants commented:
> As the sequence of the entries in the speed_constraints list matters a lot, is it better to have the 'smaller' terms at the 'left' of the list ( small indices) or towards the end of the list (large indices)?

I don't know. This is going to be problem specific and relates to the nature of the nonholonomic constraint expressions.

moorepants committed:
Merge branch 'master' into whipple

moorepants committed:
Solve symbolic motion constraints with CRAMER method, avoids divide-by-zero issues.

moorepants committed:
Use cramer_solve() to avoid divide-by-zeros.

moorepants committed:
Rename to bicycle_validation.

moorepants commented:
The main issue with this example was that we changed the constraint symbolic solver inside KanesMethod to LUsolve() (see https://github.com/sympy/sympy/issues/9641 and https://github.com/sympy/sympy/issues/24780) which causes divide-by-zero issues with this bicycle formulation. I now use the cramer_solve() which is available in SymPy >=1.13 and everything checks out to at least 11 decimal places for the Basu-Mandal comparison. We have the working example in the documentation but I will merge this given it has extensive validation tooling and may be usef for the future.

moorepants committed:
Rename so that pytest doesn't try to run these tests.

moorepants commented:
Current output:

```
In [1]: run whipple.py
+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
Deriving the primary symbolic equations of motion to check.
+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
Defining reference frames.
Defining time varying symbols.
Orienting frames.
Defining constants.
Defining position vectors.
Defining holonomic constraints.
The holonomic constraint is a function of these dynamic variables:
[q4(t), q5(t), q7(t)]
Defining kinematical differential equations.
Defining angular velocities.
Defining linear velocities.
Defining nonholonomic constraints.
The nonholonomic constraints are a function of these dynamic variables:
[q4(t), q5(t), q7(t), u3(t), u4(t), u5(t), u6(t), u7(t), u8(t)]
Defining inertia.
Defining the rigid bodies.
Defining the forces and torques.
Generating Kane's equations.
The mass matrix is a function of these dynamic variables:
[q4(t), q5(t), q7(t)]
The forcing vector is a function of these dynamic variables:
[T4(t), T6(t), T7(t), q4(t), q5(t), q7(t), u3(t), u4(t), u5(t), u6(t), u7(t), u8(t)]
Writing mass matrix and forcing vector to files.
++++++++++++++++++++++++++++++++++++++++++++++++++
Checking against Symengine high precision results.
++++++++++++++++++++++++++++++++++++++++++++++++++
Loading numerical input parameters.
id33 not added to sub dict.
if33 not added to sub dict.
ic12 not added to sub dict.
ic23 not added to sub dict.
ie12 not added to sub dict.
ie23 not added to sub dict.
id33 not added to sub dict.
if33 not added to sub dict.
ic12 not added to sub dict.
ic23 not added to sub dict.
ie12 not added to sub dict.
ie23 not added to sub dict.
Compare dependendent speeds with Symengine evalf().
u3(t): -0.78300335270650,-0.78300335270653
u5(t): 0.01191855280690,0.01191855280690
u8(t): -8.01336205841550,-8.01336205841539
Evaluating numerically with symengine
The state derivatives from high precision evaluation:
[ 7.85552811 -1.84725541  4.6198904  -0.83532817 -0.12055439 -2.45480729]
Running checks
Evaluating numerically with xreplace
Mass matrix from xreplace is correct to machine precision
Forcing vector from xreplace is correct to machine precision
Evaluating with autowrap C
Mass matrix from autowrap C is correct to machine precision
Forcing vector from autowrap C is correct to machine precision
Evaluating with autowrap Fortran
Mass matrix from autowrap Fortran is correct to machine precision
Forcing vector from autowrap Fortran is correct to machine precision
Mass matrix from CythonMatrixGenerator with cse is correct to machine precision
Forcing vector from CythonMatrixGenerator with cse is correct to machine precision
++++++++++++++++++++++++++++++++++++++++++++++++++++++++
Checking primary equations of motion versus Basu-Mandal.
++++++++++++++++++++++++++++++++++++++++++++++++++++++++
Generating a right hand side function.
Generating rhs
The state derivatives from code gen:
[-0.78300335 -0.60684258 -0.48598247  0.01191855  7.85552811 -1.84725541
  4.6198904  -0.83532817 -0.12055439 -2.45480729]
Generating output dictionary.
Comparing numerical results to the Basu-Mandal values:
Results from Symengine evaluation
Matched: xdd:
  Expected: -0.5041626315047000
    Actual: -0.5041626315048091
Matched: ydd:
  Expected: -0.3449706619454000
    Actual: -0.3449706619455042
Matched: zdd:
  Expected: -1.4604528332980000
    Actual: -1.4604528332980475
Matched: thetadd:
  Expected: 0.8353281706379000
    Actual: 0.8353281706376495
Matched: psidd:
  Expected: -7.8555281128244001
    Actual: -7.8555281128244676
Matched: phidd:
  Expected: 0.1205543897884000
    Actual: 0.1205543897886408
Matched: psifdd:
  Expected: -4.6198904039402997
    Actual: -4.6198904039393760
Matched: betardd:
  Expected: 1.8472554144216999
    Actual: 1.8472554144220372
Matched: betafdd:
  Expected: 2.4548072904549998
    Actual: 2.4548072904553715
Results from CythonODEGenerator
Matched: xdd:
  Expected: -0.5041626315047000
    Actual: -0.5041626315048090
Matched: ydd:
  Expected: -0.3449706619454000
    Actual: -0.3449706619455035
Matched: zdd:
  Expected: -1.4604528332980000
    Actual: -1.4604528332980480
Matched: thetadd:
  Expected: 0.8353281706379000
    Actual: 0.8353281706376527
Matched: psidd:
  Expected: -7.8555281128244001
    Actual: -7.8555281128244703
Matched: phidd:
  Expected: 0.1205543897884000
    Actual: 0.1205543897886393
Matched: psifdd:
  Expected: -4.6198904039402997
    Actual: -4.6198904039393520
Matched: betardd:
  Expected: 1.8472554144216999
    Actual: 1.8472554144220366
Matched: betafdd:
  Expected: 2.4548072904549998
    Actual: 2.4548072904553728
```
